File: Data Structures & Algorithms/balanced-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int height(TreeNode* root) {
        if (!root) return 0;
        return 1 + max(height(root->left), height(root->right));
    }
    bool isBalanced(TreeNode* root) {
        if (!root || (!root->right && !root->left)) return true;
        if (root->left && root->right) {
            if (isBalanced(root->left) && isBalanced(root->right)) {
                if (abs(height(root->left) - height(root->right)) <= 1) {
                    return true;
                }
                return false;
            }
        } 
        else if (root->left) {
            if (root->left->left || root->left->right) {
                return false;
            }
            else {return true;}
        }
        
        else if (root->right) {
            if (root->right->left || root->right->right) {
                return false;
            }
            else {return true;}
        }
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.m = 0
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def dfs(root):
            if not root:
                return 0
            temp = 0
            a = 0
            b = 0
            if root.left:
                temp += 1
                a += 1
            if root.right:
                temp += 1
                b += 1
            l = dfs(root.left)
            r = dfs(root.right)
            temp += l + r
            self.m = temp if temp > self.m else self.m # m = 1
            return max(a+l, b+r)
        
        res = dfs(root)
        return self.m